2021q4/vdso.adoc

```diff
@@ -0,0 +1,79 @@
+== VDSO on amd64
+
+Contact: Konstantin Belousov <kib@FreeBSD.org>  
+
+VDSO AKA Virtual Dynamic Shared Object is a shared object (in the
+common speak, dynamic library) that is injected into the executed
+image by common efforts from kernel and dynamic linker.  It does not
+exists on disk as a standalone .so, and there are no instructions in
+the ELF format that cause the injection.  It is done by system to
+implement some functionality for the consumers.
+
+FreeBSD already has a lot of features that otherwise typically done
+using VDSO (in Linux), but really not requiring that complication.
+Main reason why it is possible is the often mentioned co-evolution of
+the kernel and C runtime.  We can naturally introduce features that
+require implementation both in kernel, and support in the userspace
+parts, since FreeBSD is developed as the whole.  Surprisingly, it also
+allows for kernel and dynamic linker to know much less (and not
+enforce anything) about userspace consumers of interfaces.
+
+For instance, syscall-less wall clock was implemented long time ago,
+by kernel providing time hands blob in the shared page, and C library
+knowing about the blob location and supported algorithms.  There is no
+need for VDSO that interposes some libc symbols, or provides services
+that are named by known symbols to userspace.
```

Comment:
The 'symbols' term is crusial there

Reply:
Then I don't understand what you're trying to say because "named by symbols that userspace knows" and "have names known to userspace" mean exactly the same to me. If names aren't symbols, what are they?

Reply:
Names are names, existence of name for something does not imply anything technically.  Symbol is the real construct that is observed and handled by software.

For instance, mentioned timehands do have name, we refer to them as 'timehands', but there is no symbol for it.